### ISSUE TITLE
gmake missing dependency

### ISSUE THREAD

**Issue by mjs-arm** [premake/premake-core#1094]
premake gmake backend does not generate a dependency between the "prebuild" target and the targets associated with the actual build, ie $(TARGET).   Hence prebuild can run in parallel with or after the build. In practice this means that if prebuild is used to generate anything used by the build proper the resulting Makefile is not parallel safe.

**Comment by samsinsane** [comment]
Does the `gmake2` action also have this dependency issue?

**Comment by tdesveauxPKFX** [comment]
`gmake2` generate correct dependencies with prebuild as far as I saw.

Ultimately, the `gmake` action will be deprecated in favor of `gmake2`, so it would make sense to use it instead.
I had to port from `gmake` to `gmake2` at work when we updated to premake5 and it was mostly painless as I remember.

I'd say this is a `wontfix` if someone can confirm and label/close?

**Comment by starkos** [comment]
Is there a reason we’re not dropping gmake for gmake2? Seems like folks are already switching over.

**Comment by tdesveauxPKFX** [comment]
Well, I don't see a reason not to. I had to switch because of issues (but I don't remember what exactly, it was some time ago) and `gmake2` looks robust enough as it is to me.

I know I have some overrides in our scripts that may be of interest but I would need to look if they are still relevant. I will try to look into it this week.

Anyway, I think it make sense to drop `gmake` or at least put some kind of warning to let people know that we will go with `gmake2`.

**Comment by samsinsane** [comment]
I'm pretty happy to see `gmake` dropped in favour of `gmake2`. :+1:

**Comment by tvandijck** [comment]
Dropping 'gmake'  would also result in dropping the D module.

Which is why we chose this path in the first place. Someone needs to take the time to port the D module to gmake2 first.

**Comment by mjs-arm** [comment]
The gmake2 backend also has the same problem, there is no dependency between the prebuild and the build proper hence no guarantee that the prebuild steps complete before other the build tries to build other components.

**Comment by tdesveauxPKFX** [comment]
Are you using a c# makefile? There indeed seems to be an issue for them.
c++ makefiles have correct dependencies.

I will try to look into it tonight.

For a quick fix right now, you can add the following overrides in your project:
```
require("gmake2")

local gmake2 = premake.modules.gmake2

premake.override(gmake2, "csAllRules", function(prj, toolset)
	_p('all: $(EMBEDFILES) $(COPYFILES) $(TARGET)')
	_p('')
end)

premake.override(gmake2, "csFileRules", function(prj, toolset)
	_p('$(SOURCES): | prebuild')
end)
```
I don't have time to really test it so be there might be issues.

Also, this is under the assumption that you have a c# makefile. If you use a c++ one, the problem should lie somewhere else and it would be helpful if you could provide a reproduction case.

**Comment by mjs-arm** [comment]
The issue is with C/C++ Makefile, the attachment is a reproducer that demonstrates that the prebuild action is not complete at the point the build starts.  To recreate, unpack and run ./test.sh
[break-prebuild.tar.gz](https://github.com/premake/premake-core/files/2053740/break-prebuild.tar.gz)

**Comment by tdesveauxPKFX** [comment]
Works fine on my side the latest Premake on master.

Do you know which Premake version you are using?
I know there was some work done not that long ago on dependencies.

**Comment by mjs-arm** [comment]
$ premake5 --version
premake5 (Premake Build Script Generator) 5.0.0-alpha12

**Comment by mjs-arm** [comment]
Ok, switch to master:  e45b8cfc6b917cbc9c4f1f4cf741eab8da180521 works fine so the issue must have been resolved since alpha12.
